From 2c8d97578901d6cd4680bd186a5c523bc3e16b94 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Thu, 5 Oct 2023 14:43:20 +0200
Subject: [PATCH 0114/1010] net: dovetail: fix use-after-free in
 __sock_release()

Although related, socket and sock struct instances have distinct
lifetimes. As a consequence, the former might have been freed already
when the latter is released eventually. The wrong assumption that both
would have the same lifetime caused __sock_release() to spuriously
refer to a deallocated sock struct in a call to the sock_oob_detach()
handler (caught by Kasan).

To fix this, call sock_oob_detach() from __sk_destruct() instead,
which matches the semantics of the former for the oob side.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/net/sock.h | 11 +++++++++++
 net/core/sock.c    |  2 ++
 net/socket.c       |  9 +--------
 3 files changed, 14 insertions(+), 8 deletions(-)

diff --git a/include/net/sock.h b/include/net/sock.h
index 95ff9265eb..c4780736a8 100644
--- a/include/net/sock.h
+++ b/include/net/sock.h
@@ -1907,6 +1907,17 @@ void sock_pfree(struct sk_buff *skb);
 #define sock_edemux sock_efree
 #endif
 
+#ifdef CONFIG_NET_OOB
+static inline void sock_oob_destruct(struct sock *sk)
+{
+	void sock_oob_detach(struct sock *sk);
+	if (sk->oob_data)
+		sock_oob_detach(sk);
+}
+#else
+static inline void sock_oob_destruct(struct sock *sk) { }
+#endif
+
 int sk_setsockopt(struct sock *sk, int level, int optname,
 		  sockptr_t optval, unsigned int optlen);
 int sock_setsockopt(struct socket *sock, int level, int op,
diff --git a/net/core/sock.c b/net/core/sock.c
index d8c0650322..ab3bb39ba7 100644
--- a/net/core/sock.c
+++ b/net/core/sock.c
@@ -2138,6 +2138,8 @@ static void __sk_destruct(struct rcu_head *head)
 	struct sock *sk = container_of(head, struct sock, sk_rcu);
 	struct sk_filter *filter;
 
+	sock_oob_destruct(sk);
+
 	if (sk->sk_destruct)
 		sk->sk_destruct(sk);
 
diff --git a/net/socket.c b/net/socket.c
index 2ac68d25f7..3a41c04311 100644
--- a/net/socket.c
+++ b/net/socket.c
@@ -156,7 +156,7 @@ int __weak sock_oob_attach(struct socket *sock)
 	return 0;
 }
 
-void __weak sock_oob_detach(struct socket *sock)
+void __weak sock_oob_detach(struct sock *sk)
 {
 }
 
@@ -209,10 +209,6 @@ static inline int sock_oob_attach(struct socket *sock)
 	return 0;
 }
 
-static inline void sock_oob_detach(struct socket *sock)
-{
-}
-
 static int sock_oob_bind(struct socket *sock,
 			struct sockaddr *addr, int len)
 {
@@ -749,9 +745,6 @@ EXPORT_SYMBOL(sock_alloc);
 
 static void __sock_release(struct socket *sock, struct inode *inode)
 {
-	if (sock_oob_capable(sock))
-		sock_oob_detach(sock);
-
 	if (sock->ops) {
 		struct module *owner = sock->ops->owner;
 
-- 
2.30.2

